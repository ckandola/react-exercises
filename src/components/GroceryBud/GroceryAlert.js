import React, {useEffect} from 'react';

const GroceryAlert = () => {
    return (
        <h2>Alert</h2>
    );
};

export default GroceryAlert;
