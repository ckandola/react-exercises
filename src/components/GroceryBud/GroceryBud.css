/*
=============== 
Variables
===============
*/

:root {
    /* dark shades of primary color*/
    --clr-primary-1: hsl(205, 86%, 17%);
    --clr-primary-2: hsl(205, 77%, 27%);
    --clr-primary-3: hsl(205, 72%, 37%);
    --clr-primary-4: hsl(205, 63%, 48%);
    /* primary/main color */
    --clr-primary-5: hsl(205, 78%, 60%);
    /* lighter shades of primary color */
    --clr-primary-6: hsl(205, 89%, 70%);
    --clr-primary-7: hsl(205, 90%, 76%);
    --clr-primary-8: hsl(205, 86%, 81%);
    --clr-primary-9: hsl(205, 90%, 88%);
    --clr-primary-10: hsl(205, 100%, 96%);
    /* darkest grey - used for headings */
    --clr-grey-1: hsl(209, 61%, 16%);
    --clr-grey-2: hsl(211, 39%, 23%);
    --clr-grey-3: hsl(209, 34%, 30%);
    --clr-grey-4: hsl(209, 28%, 39%);
    /* grey used for paragraphs */
    --clr-grey-5: hsl(210, 22%, 49%);
    --clr-grey-6: hsl(209, 23%, 60%);
    --clr-grey-7: hsl(211, 27%, 70%);
    --clr-grey-8: hsl(210, 31%, 80%);
    --clr-grey-9: hsl(212, 33%, 89%);
    --clr-grey-10: hsl(210, 36%, 96%);
    --clr-white: #fff;
    --clr-red-dark: hsl(360, 67%, 44%);
    --clr-red-light: hsl(360, 71%, 66%);
    --clr-green-dark: hsl(125, 67%, 44%);
    --clr-green-light: hsl(125, 71%, 66%);
    --clr-black: #222;
    --transition: all 0.3s linear;
    --spacing: 0.1rem;
    --radius: 0.25rem;
    --light-shadow: 0 5px 15px rgba(0, 0, 0, 0.1);
    --dark-shadow: 0 5px 15px rgba(0, 0, 0, 0.2);
    --max-width: 1170px;
    --fixed-width: 620px;
  }
  
  .gList-btn {
    text-transform: uppercase;
    background: transparent;
    color: var(--clr-black);
    padding: 0.375rem 0.75rem;
    letter-spacing: var(--spacing);
    display: inline-block;
    transition: var(--transition);
    font-size: 0.875rem;
    border: 2px solid var(--clr-black);
    cursor: pointer;
    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.2);
    border-radius: var(--radius);
  }
  .gList-btn:hover {
    color: var(--clr-white);
    background: var(--clr-black);
  }
  /* section */
  .gList-section {
    padding: 5rem 0;
  }
  
  .gList-section-center {
    width: 90vw;
    margin: 0 auto;
    max-width: 35rem;
    margin-top: 8rem;
  }
  @media screen and (min-width: 992px) {
    .gList-section-center {
      width: 95vw;
    }
  }
  .gList-main {
    min-height: 100vh;
    display: grid;
    place-items: center;
  }
  /*
  =============== 
  Grocery List
  ===============
  */
  .gList-section-center {
    background: var(--clr-white);
    border-radius: var(--radius);
    box-shadow: var(--light-shadow);
    transition: var(--transition);
    padding: 2rem;
  }
  .gList-section-center:hover {
    box-shadow: var(--dark-shadow);
  }
  .gList-alert {
    margin-bottom: 1rem;
    height: 1.25rem;
    display: grid;
    align-items: center;
    text-align: center;
    font-size: 0.7rem;
    border-radius: 0.25rem;
    letter-spacing: var(--spacing);
    text-transform: capitalize;
  }
  .gList-alert-danger {
    color: #721c24;
    background: #f8d7da;
  }
  .gList-alert-success {
    color: #155724;
    background: #d4edda;
  }
  .gList-grocery-form h3 {
    color: var(--clr-primary-1);
    margin-bottom: 1.5rem;
    text-align: center;
  }
  .gList-form-control {
    display: flex;
    justify-content: center;
  }
  .gList-grocery {
    padding: 0.25rem;
    padding-left: 1rem;
    background: var(--clr-grey-10);
    border-top-left-radius: var(--radius);
    border-bottom-left-radius: var(--radius);
    border-color: transparent;
    font-size: 1rem;
    flex: 1 0 auto;
    color: var(--clr-grey-5);
  }
  .gList-grocery::placeholder {
    font-family: var(--ff-secondary);
    color: var(--clr-grey-5);
  }
  .gList-submit-btn {
    background: var(--clr-primary-8);
    border-color: transparent;
    flex: 0 0 5rem;
    display: grid;
    align-items: center;
    padding: 0.25rem;
    text-transform: capitalize;
    letter-spacing: 2px;
    border-top-right-radius: var(--radius);
    border-bottom-right-radius: var(--radius);
    cursor: pointer;
    content: var(--clr-primary-5);
    transition: var(--transition);
    font-size: 0.85rem;
  }
  .gList-submit-btn:hover {
    background: var(--clr-primary-5);
    color: var(--clr-white);
  }
  
  .gList-grocery-container {
    margin-top: 2rem;
  }
  
  .gList-grocery-item {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-bottom: 0.5rem;
    transition: var(--transition);
    padding: 0.25rem 1rem;
    border-radius: var(--radius);
    text-transform: capitalize;
  }
  .gList-grocery-item:hover {
    color: var(--clr-grey-5);
    background: var(--clr-grey-10);
  }
  .gList-grocery-item:hover .title {
    color: var(--clr-grey-5);
  }
  .gList-title {
    margin-bottom: 0;
    color: var(--clr-grey-1);
    letter-spacing: 2px;
    transition: var(--transition);
  }
  .gList-edit-btn,
  .gList-delete-btn {
    background: transparent;
    border-color: transparent;
    cursor: pointer;
    font-size: 0.7rem;
    margin: 0 0.15rem;
    transition: var(--transition);
  }
  .gList-edit-btn {
    color: var(--clr-green-light);
  }
  .gList-edit-btn:hover {
    color: var(--clr-green-dark);
  }
  .gList-delete-btn {
    color: var(--clr-red-light);
  }
  .gList-delete-btn:hover {
    color: var(--clr-red-dark);
  }
  .gList-clear-btn {
    text-transform: capitalize;
    width: 10rem;
    height: 1.5rem;
    display: grid;
    align-items: center;
    background: transparent;
    border-color: transparent;
    color: var(--clr-red-light);
    margin: 0 auto;
    font-size: 0.85rem;
    letter-spacing: var(--spacing);
    cursor: pointer;
    transition: var(--transition);
    margin-top: 1.25rem;
  }
  .gList-clear-btn:hover {
    color: var(--clr-red-dark);
  }
