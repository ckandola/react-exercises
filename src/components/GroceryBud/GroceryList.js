import React from 'react';
import { FaEdit, FaTrash } from 'react-icons/fa';

const GroceryList = ({items, removeItem, editItem}) => {
    return (
        <div className="gList-grocery-list">
            {items.map(item => {
                const {id, title} = item;
                return (
                    <article className="gList-grocery-item" key={id}>
                        <p className="gList-title">{title}</p>
                        <div className="gList-btn-container">
                            <button className="gList-edit-btn"
                                    onClick={() => editItem(id)}>
                                <FaEdit />
                            </button>
                            <button className="gList-delete-btn"
                                    onClick={() => removeItem(id)}>
                                <FaTrash />
                            </button>
                        </div>
                    </article>
                );
            })}
        </div>
    );
};

export default GroceryList;
