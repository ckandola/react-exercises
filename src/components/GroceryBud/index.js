// from https://youtu.be/a_7Z7C_JCyo?t=14077

import React, {useState} from 'react';
import './GroceryBud.css';
import GroceryList from './GroceryList';
import GroceryAlert from './GroceryAlert';

const GroceryBud = () => {
    const [name, setName] = useState('');
    const [list, setList] = useState([]);
    const [isEditing, setIsEditing] = useState(false);
    const [editID, setEditID] = useState(null);
    const [alert, setAlert] = useState({show: false, message: '', type: ''});

    const handleSubmit = (e) => {
        e.preventDefault();
    }

    return (
        <section className="gList-section-center">
            <form className="gList-grocery-form" onSubmit={handleSubmit}>
                {alert.show && <GroceryAlert />}
            </form>
            <div className="gList-grocery-container">
                <GroceryList />
                <button className="gList-clear-btn">
                    clear items
                </button>
            </div>
        </section>
    );
};

export default GroceryBud;
