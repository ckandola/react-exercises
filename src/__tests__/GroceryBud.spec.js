import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import GroceryAlert from '../components/GroceryBud/GroceryAlert';
import GroceryList from '../components/GroceryBud/GroceryList';
import GroceryBud from '../components/GroceryBud';

describe('GroceryBud tests', () => {
    
    const wrapper = mount(<GroceryBud />);

    it('matches the snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('adds `Meat` to list when typed into input and Submit is clicked', () => {
        wrapper.find('input').simulate('change', {target: {value: 'Meat'}});
        wrapper.find('form').simulate('submit', {preventDefault () {} });
        expect(wrapper.find('.gList-title').getElement().props.children).toBe('Meat');
    });

    it('changes `Meat` to `Yogurt` when its edit button is clicked, `Yogurt` is typed into the input, and Submit is clicked', () => {
        wrapper.find('.gList-edit-btn').simulate('click');
        wrapper.find('input').simulate('change', {target: {value: 'Yogurt'}});
        wrapper.find('form').simulate('submit', {preventDefault () {}});
        expect(wrapper.find('.gList-title').getElement().props.children).toBe('Yogurt');
    });

    it('removes `Yogurt` from the list when its remove button is clicked', () => {
        wrapper.find('.gList-delete-btn').simulate('click');
        expect(wrapper.find('gList-title')).toHaveLength(0);
    });

    it('clears list of all items when Clear Items button is clicked', () => {
        wrapper.find('input').simulate('change', {target: {value: 'A'}});
        wrapper.find('form').simulate('submit', {preventDefault () {}});
        wrapper.find('input').simulate('change', {target: {value: 'B'}});
        wrapper.find('form').simulate('submit', {preventDefault () {}});
        expect(wrapper.find('.gList-title')).toHaveLength(2);
        wrapper.find('.gList-clear-btn').simulate('click');
        expect(wrapper.find('.gList-title')).toHaveLength(0);
    });

    describe('GroceryAlert tests', () => {
        
        const mockAlert = jest.fn();

        it('matches the snapshot', () => {
            const wrapper = shallow(<GroceryAlert type={"success"} msg={"Tested successfully"} removeAlert={mockAlert} />);
            expect(toJson(wrapper)).toMatchSnapshot();
            wrapper.unmount();
        });
        
        it('calls `removeAlert` function after 3 seconds', () => {
            jest.useFakeTimers();
            const wrapper = shallow(<GroceryAlert type={"success"} msg={"Tested successfully again"} removeAlert={mockAlert} />);
            expect(mockAlert.mock.calls.length).toBe(0)
            jest.advanceTimersByTime(3000);
            expect(mockAlert.mock.calls.length).toBe(1);
            jest.useRealTimers();
            wrapper.unmount();
        });

        it('renders class for alerts depending on the `type` property', () => {
            const wrapperSuccess = shallow(<GroceryAlert type={"success"} msg={"This is a success"} removeAlert={mockAlert} />);
            const wrapperDanger = shallow(<GroceryAlert type={"danger"} msg={"This is dangerous"} removeAlert={mockAlert} />);
            expect(wrapperSuccess.find('.gList-alert-success').length).toBe(1);
            expect(wrapperDanger.find('.gList-alert-danger').length).toBe(1);
        });
    });

    describe('GroceryList tests', () => {

        const mockList = [
            {id: 0, title: "test"},
        ];
        
        const mockRemove = jest.fn();
        const mockEdit = jest.fn();
        
        const wrapper = shallow(<GroceryList items={mockList} removeItem={mockRemove} editItem={mockEdit} />);

        it('matches the snapshot', () => {
            expect(toJson(wrapper)).toMatchSnapshot();
        });
        
        it('calls `editItem` when edit button is pressed', () => {
            wrapper.find('.gList-edit-btn').simulate('click');
            expect(mockEdit.mock.calls.length).toBe(1);
        });
        
        it('calls `removeItem` when remove button is pressed', () => {
            wrapper.find('.gList-delete-btn').simulate('click');
            expect(mockRemove.mock.calls.length).toBe(1);
        });
    });

});
