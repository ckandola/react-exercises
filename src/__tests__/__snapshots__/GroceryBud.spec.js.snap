// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroceryBud tests GroceryAlert tests matches the snapshot 1`] = `
<p
  className="gList-alert gList-alert-success"
>
  Tested successfully
</p>
`;

exports[`GroceryBud tests GroceryList tests matches the snapshot 1`] = `
<div
  className="gList-grocery-list"
>
  <article
    className="gList-grocery-item"
    key="0"
  >
    <p
      className="gList-title"
    >
      test
    </p>
    <div
      className="gList-btn-container"
    >
      <button
        className="gList-edit-btn"
        onClick={[Function]}
      >
        <FaEdit />
      </button>
      <button
        className="gList-delete-btn"
        onClick={[Function]}
      >
        <FaTrash />
      </button>
    </div>
  </article>
</div>
`;

exports[`GroceryBud tests matches the snapshot 1`] = `
<GroceryBud>
  <section
    className="gList-section-center"
  >
    <form
      className="gList-grocery-form"
      onSubmit={[Function]}
    >
      <h3>
        Grocery Bud
      </h3>
      <div
        className="gList-form-control"
      >
        <input
          className="gList-grocery"
          onChange={[Function]}
          placeholder="e.g. milk"
          type="text"
          value=""
        />
        <button
          className="gList-submit-btn"
          onClick={[Function]}
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </section>
</GroceryBud>
`;
